aten/src/ATen/native/cuda/MultiTensorApply.cuh

```diff
@@ -38,13 +49,213 @@ __device__ __forceinline__ void load_store(
+// vectors into a device buffer allocated with CUDACachingAllocator. Thanks to
+// the allocator's stream awareness, the buffer can be safely accessed by the
+// kernel as long as the tensor owning the buffer is alive at the time of the
+// kernel launch. This tensor is returned to the caller, who is responsible for
+// keeping it alive until the kernel is launched.
+template <int n, typename... Vectors>
+std::tuple<DevArrayPack, c10::optional<at::Tensor>> pack_vectors(
+    const at::Device& device,
+    std::vector<const void*> (&addresses)[n],
+    const Vectors&... vecs) {
+  std::vector<const void*> ptrs;
+  std::vector<size_t> sizes;
+  std::vector<size_t> offsets;
+  size_t total_bytes = 0;
+
+  for (auto d = 0; d < n; ++d) {
+    pack_vectors_helper(ptrs, sizes, offsets, total_bytes, addresses[d]);
+  }
+  (pack_vectors_helper(ptrs, sizes, offsets, total_bytes, vecs), ...);
+
+  TORCH_CHECK(ptrs.size() <= DevArrayPack::max_arrays);
+  DevArrayPack pack{};
+
+  // Use the small buffer in DevArrayPack to pack the vectors
+  if (total_bytes < DevArrayPack::small_buffer_size) {
+    pack.buffer_ptr = nullptr;
+    for (const auto i : c10::irange(ptrs.size())) {
+      pack.offsets[i] = offsets[i];
+      memcpy(pack.small_buffer + offsets[i], ptrs[i], sizes[i]);
+    }
+    return std::make_tuple(pack, c10::optional<at::Tensor>(c10::nullopt));
+  }
+
+  bool is_capturing = at::cuda::currentStreamCaptureStatusMayInitCtx() !=
+      at::cuda::CaptureStatus::None;
+
+  // Use dynamically allocated buffer to pack the vectors
+  // First, prepare a host buffer to stage the data to be copied to device
+  at::Tensor buf_tensor;
+  if (!is_capturing) {
+    // No special treatment is needed when not capturing, as the as the buffer
+    // is allocated and populated on every invocation.
+    buf_tensor = at::empty(
+        {static_cast<int64_t>(total_bytes)},
+        // Only use pinned memory when not capturing
+        at::TensorOptions().dtype(at::kByte).pinned_memory(true));
+  } else {
+    // With CUDA Graph, we need the data to have the same lifetime as the
+    // capturing graph because the data is an extension of the launch argument.
+    // This is achieved by dynamically allocating a buffer, managing the
+    // lifetime of the buffer with a CUDA User Object, and transferring the
+    // ownership of the CUDA User Object to the capturing graph.
+    //
+    // NOTE: based on PR feedback, we want to be conservative and replicate the
+    // behavior w/o CUDA Graph first. This means that a HtoD copy will be
+    // issued on every replay. This copy can be eliminated by having the CUDA
+    // User Object manage a device buffer.
+    uint8_t* buf = new uint8_t[total_bytes];
+
+    // Manage the ownership of buf with a cuda user object
+    cudaUserObject_t user_object;
+    C10_CUDA_CHECK(cudaUserObjectCreate(
+        &user_object,
+        buf,
+        [](void* buf) { delete[] reinterpret_cast<uint8_t*>(buf); },
```

Comment:
The raw "new" is not the best... Can you not capture the Tensor by value in this lambda to keep it alive until the lambda is called and them reset the shared_ptr in the lambda to free it? :D

Reply:
Yeah agreed. However, `cudaUserObjectCreate` takes a raw function pointer of type `void(CUDART_CB* cudaHostFn_t )( void*  userData )` which means capturing a tensor ref isn't possible.